app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt

```diff
@@ -0,0 +1,110 @@
+package compiler.intermediate_form
+
+import compiler.common.reference_collections.ReferenceMap
+import compiler.common.reference_collections.ReferenceSet
+import compiler.common.reference_collections.referenceHashMapOf
+import compiler.common.reference_collections.referenceHashSetOf
+import compiler.common.reference_collections.referenceMapOf
+
+abstract class DataFlowAnalyser<SL> {
+    abstract val backward: Boolean
+
+    abstract val entryPointValue: SL
+
+    abstract val latticeMaxElement: SL
+
+    abstract fun latticeMeetFunction(elements: Collection<SL>): SL
+
+    abstract fun transferFunction(instruction: Instruction, inValue: SL): SL
+
+    private fun getPredecessors(linearProgram: List<Asmable>): ReferenceMap<Instruction, List<Instruction>> {
+        val instructionList = linearProgram.filterIsInstance<Instruction>()
+
+        // calculate label to instruction translation
+        val labelToInstruction = mutableMapOf<String, Instruction>()
+        for ((index, asmable) in linearProgram.withIndex().reversed())
+            if (asmable is Label)
+                labelToInstruction[asmable.label] = linearProgram[index + 1].let {
+                    when (it) {
+                        is Label -> labelToInstruction[it.label]!!
+                        is Instruction -> it
+                    }
+                }
+
+        // calculate the predecessors map (forward direction)
+        val predecessors = referenceHashMapOf<Instruction, MutableList<Instruction>>()
+        instructionList.forEach { predecessors[it] = mutableListOf() }
+        fun addPredecessor(from: Instruction, to: Instruction) {
+            if (backward) predecessors[from]!!.add(to)
+            else predecessors[to]!!.add(from)
+        }
+
+        instructionList.forEachIndexed { index, it ->
+            when (it) {
+                is Instruction.ConditionalJumpInstruction -> {
+                    addPredecessor(it, labelToInstruction[it.targetLabel]!!)
+                    addPredecessor(it, instructionList[index + 1])
+                }
+
+                is Instruction.UnconditionalJumpInstruction ->
+                    addPredecessor(it, labelToInstruction[it.targetLabel]!!)
+
+                is Instruction.RetInstruction -> {}
+
+                is Instruction.InPlaceInstruction ->
+                    addPredecessor(it, instructionList[index + 1])
+            }
+        }
+
+        return referenceMapOf(predecessors.map { it.toPair() })
+    }
+
+    private fun getEntryPoints(instructionList: List<Instruction>): ReferenceSet<Instruction> {
+        return if (backward)
+            referenceHashSetOf(instructionList.filterIsInstance<Instruction.RetInstruction>())
+        else
+            referenceHashSetOf(instructionList.first())
```

Comment:
Is `instructionList` empty for the empty program?

Reply:
I assume that every execution path must lead to RET instruction.